### PR TITLE
let cached provider return bytes

### PR DESCRIPTION
**Pull request by danielnorberg**
Change the `Provider::get()` return type to an enum `BlobRead` which can either hold `Bytes` or `ReadStream` so that users can get bytes from the cached provider without copying.

Note that this is a backwards incompatible API change.

Also changes the cached provider to read into a pre-sized vec to avoid expensive realloc.

Very much open to suggestions for better or more rustilicious ways to achieve this.

### CODE REVIEW

File: blobnet/src/lib.rs [hunk 1]
```diff
@@ -143,6 +145,36 @@ type BlobRange = Option<(u64, u64)>;
...
+/// A buffer or stream of bytes from some data source.
+pub enum BlobRead<'a> {
+    /// A buffer of bytes
+    Bytes(Bytes),
+
+    /// A stream of bytes
+    Stream(ReadStream<'a>),
+}
+
+// TODO(dano): could this be impl From or Into instead?
```
Comment (thundergolfer) [inline blobnet/src/lib.rs:157]: Did you run into tricky compile issues trying `From` trait impls?

**Review by thundergolfer — APPROVED** [review]
LGTM, nice one. 

I think the API change is fine, we just got to got to v0.4.0 when doing the crates.io release.

File: blobnet/src/lib.rs (hunk 1 above)
Comment (danielnorberg) [reply blobnet/src/lib.rs:157]: Yeah, the From impl for Bytes and AsyncRead conflicted 😐

Comment (danielnorberg) [reply blobnet/src/lib.rs:157]: Screenshot 2023-04-05 at 08 44 43

Comment (danielnorberg) [reply blobnet/src/lib.rs:157]:
```rust
impl<'a> From<Bytes> for BlobRead<'a> {
    fn from(value: Bytes) -> Self {
        BlobRead::Bytes(value)
    }
}

impl<'a, T: AsyncRead + Send + 'a> From<T> for BlobRead<'a> {
    fn from(value: T) -> Self {
        BlobRead::Stream(Box::pin(value))
    }
}
```

yields

```
error[E0119]: conflicting implementations of trait `From<bytes::Bytes>` for type `BlobRead<'_>`
   --> blobnet/src/lib.rs:176:1
    |
170 | impl<'a> From<Bytes> for BlobRead<'a> {
    | ------------------------------------- first implementation here
...
176 | impl<'a, T: AsyncRead + Send + 'a> From<T> for BlobRead<'a> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `BlobRead<'_>`
    |
    = note: upstream crates may add a new impl of trait `tokio::io::AsyncRead` for type `bytes::Bytes` in future versions
```

**Pull request merged**